core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -179,13 +181,18 @@ public ServerImpl shutdown() {
     return this;
   }
 
-  /**
-   * NOT YET IMPLEMENTED. This method currently behaves identically to shutdown().
-   */
-  // TODO(ejona86): cancel preexisting calls.
   @Override
   public ServerImpl shutdownNow() {
     shutdown();
+    Collection<ServerTransport> transportsCopy;
+    Status nowStatus = Status.UNAVAILABLE.withDescription("Server shutdownNow invoked");
+    synchronized (lock) {
+      shutdownNowStatus = nowStatus;
+      transportsCopy = new ArrayList<ServerTransport>(transports);
+    }
+    for (ServerTransport transport : transportsCopy) {
+      transport.shutdownNow(nowStatus);
```

Comment:
Ah, I see the cause for confusion now. I need the one in `serverShutdown()` because there is a race of a new transport being created between requesting the server to stop accepting connections and it actually doing so. I need the one in `shutdownNow()` because `serverShutdown()` may have already been called due to a previous call to `shutdown()`. There's a strong need for some comments here (I didn't previously think about these two sets of changes as a whole).